src/task_manager.cc

```diff
@@ -143,16 +154,89 @@ namespace detail {
 
 		if(auto cgid = tsk->get_collective_group_id(); cgid != 0) {
 			if(auto prev = last_collective_tasks.find(cgid); prev != last_collective_tasks.end()) {
-				tsk->add_dependency({task_map.at(prev->second).get(), dependency_kind::ORDER_DEP});
+				add_dependency(tsk.get(), task_map.at(prev->second).get(), dependency_kind::ORDER_DEP);
 				last_collective_tasks.erase(prev);
 			}
 			last_collective_tasks.emplace(cgid, tid);
 		}
 	}
 
-	void task_manager::invoke_callbacks(task_id tid) {
+	task_id task_manager::get_new_tid() { return next_task_id++; }
```

Comment:
It would probably make sense to move this into the header to allow the compiler to inline this. This would also be more consistent with other very simple one statement functions that are already defined in the header.